### DISCUSSION TITLE
RegEx manager for updaing configuration presets

### DISCUSSION THREAD

**Discussion by electriquo** [renovatebot/renovate#16696] [Request Help] [answered]
After reading https://github.com/renovatebot/renovate/issues/16679#issuecomment-1191461860 I thought to try the regex manager for that until Renovate will introduce a builtin support for updating its own configuration presets.

I have constructed the following regex and [tested it](https://regex101.com/r/0dfzsT/1)

```regex
/(.+?[^"])"(?<provider>git(lab|hub|ea))>(?<depName>.+?)(?<path>\/\/.+?(?=#))?#(?<currentValue>.+?[^"])"/gm
```

I consulted the regex manager docs https://github.com/renovatebot/renovate/blob/497b5de6c2cd862ec25d4322342d1b6e8896ca2f/lib/modules/manager/regex/readme.md?plain=1#L50-L52 how to escape it (I used regex101 (javascript) code generator) and created the following regex configuration

```json
{
  ...
  "regexManagers": [
    {
      "fileMatch": ["(^|/).?renovate(?:rc)?(?:\\.json5?)?$"],
      "matchStrings": [
        "(.+?[^\"])\"(?<platform>git(lab|hub|ea))>(?<depName>.+?)(?<path>\\/\\/.+?(?=#))?#(?<currentValue>.+?[^\"])\""
      ],
      "datasourceTemplate": "git-tags"
    }
  ]
  ...
}
```

But I get the following error from Renovate

> `Invalid regExp for regexManagers: '(.+?[^"])"(?<platform>git(lab|hub|ea))>(?<depName>.+?)(?<path>\/\/.+?(?=#))?#(?<currentValue>.+?[^"])"'`

Also verified that the escaped regex is OK with Javascript

```javascript
# test.js
const matchString = "(.+?[^\"])\"(?<platform>git(lab|hub|ea))>(?<depName>.+?)(?<path>\\/\\/.+?(?=#))?#(?<currentValue>.+?[^\"])\"";
const re = new RegExp(matchString);
const str = JSON.stringify({
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base", "gitlab>abc/foo//path/xyz#1.5.4"]
});
const res = str.match(re);
console.log(res.groups);
``` 

The execution of the snippet outputs

```shell
[Object: null prototype] {
  platform: 'gitlab',
  depName: 'abc/foo',
  path: '//path/xyz',
  currentValue: '1.5.4'
}
```

**1. Could you help me in figuring out what is wrong with the regex?**

---

As a side note, I did not know how to utilize the [Handlebars helpers](https://docs.renovatebot.com/templates/#additional-handlebars-helpers) to use the `platform` and `depName` regex named group in the `depNameTemplate` to construct the full repository URL. E.g., `https://${platform}.com/${depName}` for the `git-tags` [datasource](https://github.com/renovatebot/renovate/tree/main/lib/modules/datasource) to work correctly.

**2. Could you help me figure out how to use construct the `depNameTemplate` based on the `platform` and `depName` regex named group for `git-tags` datasource to work?**

---

**Comment by viceice** [comment, marked as answer]
we are using RE2, so no lookahead and lookbehind support

https://github.com/uhop/node-re2#limitations-things-re2-does-not-support

**Reply by electriquo** [reply]
Thanks for the lead but I would appreciate some help.
If you would relate to my questions it will be fine and dandy.

Have to ask, what is the motivation for using a regex library which does not fully support regex, isn't it an absurd?
I also think that the RE2 limitations should be noted in Renovate docs.

**Reply by rarkins** [reply]
Warning: rudeness will not be tolerated

**Reply by electriquo** [reply]
@rarkins Did not mean to be rude in any way, I think you got me wrong. Anyhow, my **_SINCERE APOLOGIES_**.
Was only asked about the motivation of using RE2 to know whether it can be replaced.

**Reply by electriquo** [reply]
@viceice Thanks again for the lead. I've created #16709 to reflect the limitations into the docs.

**Reply by rarkins** [reply]
https://github.com/google/re2/wiki/WhyRE2

**Comment by electriquo** [comment]
@rarkins Maybe you could assist and if we would come up with a solution, I would be happy to [open a pull-request](https://github.com/renovatebot/renovate/issues/16679#issuecomment-1192197041)
